keps/sig-apps/4040-controller-name/README.md

```diff
@@ -0,0 +1,367 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature unblocks custom handling for selected standard
+Kubernetes API objects, while still allowing the default
+behavior for other API objects. For now it focuses only on Job and HPA, but other controllers can reuse this pattern.
+
+## Motivation
+
+Currently Kubernetes users, who need custom behavior for
+some API objects, can either:
+
+* disable the corresponding controller in controller-manager
+and implement a new one that provides both modified(for marked
+objects) and original(for all other) functionality.
+
+* duplicate/fork the API as CRD.
+
+However, on managed environments (like, for example, GKE),
+only the second option is available as cloud providers often
+don't allow to modify `controller-manager` configuration.
+
+### Goals
+
+* Establish a simple and common method to mark an API object as
+to be handled by a non default controller, in a similar fashion
+as pods can specify their custom scheduler.
+
+* Apply it and provide implementation for Job and HPA.
+
+### Non-Goals
+
+* Add controller name to all possible K8S objects at once.
+
+
+* While the KEP enables providing custom handling for standard
+K8S APIs,
+it doesn't focus on any customization or additional
+configuration that might be possible using it. Additional
+configuration options may be subject for the following
+enhancement proposals.
+
+
+## Proposal
+
+Add `controllerName` at the top level of Spec of API objects to
```

Comment:
in the endpointslice controller we are using a label `"endpointslice.kubernetes.io/managed-by"`

https://github.com/kubernetes/kubernetes/blob/18d05b646d09b2971dc5400bc288062b0414e8cf/staging/src/k8s.io/api/discovery/v1/well_known_labels.go#L27

I see it in alternatives but is there any special reason why label were discarded?

labels also have the benefit that are already supported by informers to filter information, that makes it easier to support environments with multiple controllers

@thockin should we try to be consistent across all the APIs about this feature? maybe is more a sig-arch topic?

Reply:
It sounded like a quite important piece of API so I went with a dedicated field, but the whole idea would also work with a label.